libp2p/connmanager.nim

```diff
@@ -0,0 +1,276 @@
+
+  try:
+    await lock.acquire()
+    trace "cleaning up connection for peer", peer = $peerInfo
+    if conn in c.muxed:
+      let muxerHolder = c.muxed[conn]
+      c.muxed.del(conn)
+
+      await muxerHolder.muxer.close()
+      if not(isNil(muxerHolder.handle)):
+        await muxerHolder.handle
+
+    if peerInfo.peerId in c.conns:
+      c.conns[peerInfo.peerId].excl(conn)
+
+      if c.conns[peerInfo.peerId].len == 0:
+        c.conns.del(peerInfo.peerId)
+
+    if not(conn.peerInfo.isClosed()):
+      conn.peerInfo.close()
+
+  finally:
+    await conn.close()
+    libp2p_peers.set(c.conns.len.int64)
+
+    if lock.locked():
+      lock.release()
+
+    trace "connection cleaned up"
+
+proc onClose(c: ConnManager, conn: Connection) {.async.} =
+  ## connection close even handler
+  ##
+  ## triggers the connections resource cleanup
+  ##
+
+  await conn.closeEvent.wait()
+  trace "triggering connection cleanup"
+  await c.cleanupConn(conn)
+
+proc selectConn*(c: ConnManager,
+                peerInfo: PeerInfo,
+                dir: Direction): Connection =
+  ## Select a connection for the provided peer and direction
+  ##
+
+  if isNil(peerInfo):
+    return
+
+  let conns = toSeq(
+    c.conns.getOrDefault(peerInfo.peerId))
+    .filterIt( it.dir == dir )
+
+  if conns.len > 0:
+    return conns[0]
+
+proc selectConn*(c: ConnManager, peerInfo: PeerInfo): Connection =
+  ## Select a connection for the provided giving priority
+  ## to outgoing connections
+  ##
+
+  if isNil(peerInfo):
+    return
+
+  var conn = c.selectConn(peerInfo, Direction.Out)
+  if isNil(conn):
+    conn = c.selectConn(peerInfo, Direction.In)
+
+  return conn
+
+proc selectMuxer*(c: ConnManager, conn: Connection): Muxer =
+  ## select the muxer for the provided connection
+  ##
+
+  if isNil(conn):
+    return
+
+  if conn in c.muxed:
+    return c.muxed[conn].muxer
+
+proc storeConn*(c: ConnManager, conn: Connection) =
+  ## store a connection
+  ##
+
+  if isNil(conn):
+    raise newException(CatchableError, "connection cannot be nil")
+
+  if isNil(conn.peerInfo):
+    raise newException(CatchableError, "empty peer info")
+
+  let peerInfo = conn.peerInfo
+  if c.conns.getOrDefault(peerInfo.peerId).len > c.maxConns:
+    trace "too many connections", peer = $conn.peerInfo,
+                                  conns = c.conns
+                                  .getOrDefault(peerInfo.peerId).len
+
+    raise newTooManyConnections()
+
+  if peerInfo.peerId notin c.conns:
+    c.conns[peerInfo.peerId] = initHashSet[Connection]()
+
+  c.conns[peerInfo.peerId].incl(conn)
+
+  # launch on close listener
+  asyncCheck c.onClose(conn)
+  libp2p_peers.set(c.conns.len.int64)
+
+proc storeOutgoing*(c: ConnManager, conn: Connection) =
+  conn.dir = Direction.Out
+  c.storeConn(conn)
+
+proc storeIncoming*(c: ConnManager, conn: Connection) =
+  conn.dir = Direction.In
+  c.storeConn(conn)
+
+proc storeMuxer*(c: ConnManager,
```

Comment:
not async, chance to use result/raises, consumer can just call `tryGet` and magically can throw

Reply:
sure, we can try with results.